### ISSUE TITLE
Weird error when running Kivy on PYPY 5.8.0

### ISSUE THREAD

**Issue by jabdoa2** [kivy/kivy#5380]
### Versions

* Python: pypy 5.8.0
* OS: Ubuntu 16.04
* Kivy: 1.10
* Kivy installation method: pip

### Description

Kivy compiles fine with pypy 5.8.0. However, I get error when running it. Can this be fixed or is that a more general issue?

### Code and Logs

```
bin/pypy3 -m kivy.uix.videoplayer
[INFO   ] [Logger      ] Record log in /home/jan/.kivy/None/kivy_17-09-13_1.txt
[INFO   ] [Kivy        ] v1.10.0
[INFO   ] [Python      ] v3.5.3 (a39af0be3a22, Jun 05 2017, 20:18:00)
[PyPy 5.8.0-beta0 with GCC 6.2.0 20160901]
[INFO   ] [Factory     ] 194 symbols loaded
 Traceback (most recent call last):
   File "/home/jan/bin/pypy3-v5.8.0-linux64/lib-python/3/runpy.py", line 193, in _run_module_as_main
     "__main__", mod_spec)
   File "/home/jan/bin/pypy3-v5.8.0-linux64/lib-python/3/runpy.py", line 85, in _run_code
     exec(code, run_globals)
   File "/home/jan/bin/pypy3-v5.8.0-linux64/site-packages/kivy/uix/videoplayer.py", line 87, in <module>
     from kivy.uix.gridlayout import GridLayout
   File "/home/jan/bin/pypy3-v5.8.0-linux64/site-packages/kivy/uix/gridlayout.py", line 93, in <module>
     from kivy.uix.layout import Layout
   File "/home/jan/bin/pypy3-v5.8.0-linux64/site-packages/kivy/uix/layout.py", line 59, in <module>
     from kivy.uix.widget import Widget
   File "/home/jan/bin/pypy3-v5.8.0-linux64/site-packages/kivy/uix/widget.py", line 244, in <module>
     from kivy.graphics import (
   File "/home/jan/bin/pypy3-v5.8.0-linux64/site-packages/kivy/graphics/__init__.py", line 89, in <module>
     from kivy.graphics.instructions import Callback, Canvas, CanvasBase, \
   File "kivy/graphics/vbo.pxd", line 7, in init kivy.graphics.instructions
     cdef class VBO:
   File "kivy/graphics/compiler.pxd", line 1, in init kivy.graphics.vbo
     cdef class GraphicsCompiler
   File "kivy/graphics/shader.pxd", line 5, in init kivy.graphics.compiler
     cdef class ShaderSource:
   File "kivy/graphics/texture.pxd", line 3, in init kivy.graphics.shader
     cdef class Texture:
   File "array.pxd", line 58, in init kivy.graphics.texture
 ValueError: array.array has the wrong size, try recompiling. Expected 24, got 72

```

**Comment by KeyWeeUsr** [comment]
I'm not aware that Kivy can run on PyPy. I'm quite curious for the compilation log, but let's go with it anyway. Up to the `Factory` part it's +/- a basic Python code if I remember correctly. After that the providers come i.e.:

* Image
* Text
* Window
  * and its GL related logs

From the log you can see the Cython code working quite well until it gets to the CPython `array.array` part. I'm not sure whether it's just the initializing phase or if it already can load data from file via image provider (I kind of doubt it though), but I thought it was because of the `pbuffer` in the `Texture`. From what I could trace by hand (I don't have pypy), I got to `pbuffer` non-related lines that apparently work **only** with CPython `array`:

* https://github.com/kivy/kivy/blob/4fcdd00a7db836d5aeb4efcbc9546930b102b8b7/kivy/graphics/texture.pyx#L647
* https://github.com/kivy/kivy/blob/082c4544415ab6bd976d9dda9c342bae109e7cbb/kivy/graphics/context.pyx#L317
* https://github.com/kivy/kivy/blob/082c4544415ab6bd976d9dda9c342bae109e7cbb/kivy/graphics/context.pyx#L49
* https://github.com/kivy/kivy/blob/082c4544415ab6bd976d9dda9c342bae109e7cbb/kivy/graphics/context.pyx#L54
* https://github.com/kivy/kivy/blob/082c4544415ab6bd976d9dda9c342bae109e7cbb/kivy/graphics/context.pyx#L19

If you want better result, just run it with `trace` option enabled and it might give you specific line it fails on.

Those lines I found, however, led me to a very similar problem in [`compreffor`](https://github.com/googlei18n/compreffor/commit/21beae578448f2762db88fe637468f974a35f857). If **only that** is the problem, we could have an `assert` statement or perhaps a check in the `setup.py` that checks if the `array` size is the correct one. 

I'm not experienced what sizes should it have though or whether we can even check it *properly* - not just dummy asserts for a close enough value. Perhaps we could make own file handling the `array` part and `(c)import` from that where we could handle the sizes. OR we could just check if the interpreter is PyPy and return a `list` bluntly if it wouldn't go the proper way. I guess it'll affect the performance quite a lot.

@tito @matham any ideas about this one?

**Comment by jabdoa2** [comment]
@KeyWeeUsr This is not a compilation problem. Everything compiles fine. Its a runtime error. I installed it using `pypy -m pip install -e .` directly on the master branch. Works fine without any output.

I retried this with pypy 6.0. Still installs fine but the error is mostly the same:

```
kivy: Factory: 194 symbols loaded
/home/jan/bin/pypy3-v6.0.0-linux64/lib-python/3/importlib/_bootstrap.py:223: UserWarning: builtins.type size changed, may indicate binary incompatibility
  return f(*args, **kwds)
Traceback (most recent call last):
  File "/home/jan/bin/pypy3-v6.0.0-linux64/lib-python/3/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/jan/bin/pypy3-v6.0.0-linux64/lib-python/3/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/data/home/jan/cloud/flipper/src/mpf/mpf/__main__.py", line 15, in <module>
    main()
  File "/data/home/jan/cloud/flipper/src/mpf/mpf/__main__.py", line 11, in main
    mpf.commands.run_from_command_line(args)
  File "/data/home/jan/cloud/flipper/src/mpf/mpf/commands/__init__.py", line 176, in run_from_command_line
    CommandLineUtility(path).execute()
  File "/data/home/jan/cloud/flipper/src/mpf/mpf/commands/__init__.py", line 152, in execute
    *self.parse_args())
  File "/data/home/jan/cloud/flipper/src/mpf-mc/mpfmc/commands/mc.py", line 197, in __init__
    from mpfmc.core.mc import MpfMc
  File "/data/home/jan/cloud/flipper/src/mpf-mc/mpfmc/core/mc.py", line 14, in <module>
    from kivy.app import App
  File "/data/home/jan/cloud/flipper/src/kivy/kivy/app.py", line 327, in <module>
    from kivy.uix.widget import Widget
  File "/data/home/jan/cloud/flipper/src/kivy/kivy/uix/widget.py", line 244, in <module>
    from kivy.graphics import (
  File "/data/home/jan/cloud/flipper/src/kivy/kivy/graphics/__init__.py", line 89, in <module>
    from kivy.graphics.instructions import Callback, Canvas, CanvasBase, \
  File "kivy/graphics/vbo.pxd", line 7, in init kivy.graphics.instructions (/data/home/jan/cloud/flipper/src/kivy/kivy/graphics/instructions.c:16520)
    cdef class VBO:
  File "kivy/graphics/compiler.pxd", line 1, in init kivy.graphics.vbo (/data/home/jan/cloud/flipper/src/kivy/kivy/graphics/vbo.c:6795)
    cdef class GraphicsCompiler
  File "kivy/graphics/shader.pxd", line 5, in init kivy.graphics.compiler (/data/home/jan/cloud/flipper/src/kivy/kivy/graphics/compiler.c:4299)
    cdef class ShaderSource:
  File "kivy/graphics/texture.pxd", line 3, in init kivy.graphics.shader (/data/home/jan/cloud/flipper/src/kivy/kivy/graphics/shader.c:13435)
    cdef class Texture:
  File "array.pxd", line 58, in init kivy.graphics.texture (/data/home/jan/cloud/flipper/src/kivy/kivy/graphics/texture.c:33076)
ValueError: array.array has the wrong size, try recompiling
```

I looked into the error and it looks like this is something in cython/pypy without a proper solution. Not sure what the best way to solve this would be. I would be happy to help but I don't know how.

**Comment by frog-o** [comment]
>I looked into the error and it looks like this is something in cython/pypy without a proper solution. Not sure what the best way to solve this would be. I would be happy to help but I don't know how.

I **might** know how 
I found [this](https://github.com/h5py/h5py/issues/1514) on h5py project on github while doing a google search for the error.  
**It help me** (it explain what going on in some what in layman terms)

I rely don't have enough knowledge on python and kivy to know 4 sure but it sound simple to me 
but  **I don't know what i am doing** and I wanted to post what I found out.
I thought it **might** be useful to you and want to know your thoughts 

from the [link above](https://github.com/h5py/h5py/issues/1514) they said it was because of the (non-portable)array.array need to be changed to bytearray would you be open to such a change?

Would this not be a proper fix?

**Comment by jabdoa2** [comment]
> from the [link above](https://github.com/h5py/h5py/issues/1514) they said it was because of the (non-portable)array.array need to be changed to bytearray would you be open to such a change?
> 
> Would this not be a proper fix?

That would definitely be worth a try. Should be easy to verify as this happens in the examples already.

**Comment by frog-o** [comment]
>from the link above they said it was because of the (non-portable)array.array need to be changed to bytearray would you be >open to such a change?

>Would this not be a proper fix?
I meant to ask one of the developers sorry ,Like I said I don't know what i am doing.

**Comment by jabdoa2** [comment]
I tried this today. I removed all `from array import array` and `from cpython.array cimport array`. Did not change the error a bit unfortunately. Not sure why.

**Comment by Julian-O** [comment]
Pypy isn't a supported "platform", so any changes here will pretty much need to come with PRs.

We don't seem to have a working proposal. 

I am going to close as Not Planned, but if someone wants to tackle Pypy support, let us know.
